cpp/include/cudf/table/table_device_view.cuh

```diff
@@ -109,4 +110,55 @@ class mutable_table_device_view
   {
   }
 };
+
+template <typename ColumnDeviceView, typename HostTableView>
+auto contiguous_copy_column_device_views(HostTableView source_view, cudaStream_t stream)
+{
+  //
+  // First calculate the size of memory needed to hold the
+  // table's ColumnDeviceViews. This is done by calling extent()
+  // for each of the table's ColumnViews columns.
+  std::size_t views_size_bytes =
+    std::accumulate(source_view.begin(), source_view.end(), 0, [](std::size_t init, auto col) {
+      return init + ColumnDeviceView::extent(col);
+    });
+  auto num_columns = std::distance(source_view.begin(), source_view.end());
+  // A buffer of CPU memory is allocated to hold the ColumnDeviceView
+  // objects. Once filled, the CPU memory is then copied to device memory
+  // and the pointer is set in the _columns member.
+  std::vector<int8_t> h_buffer(views_size_bytes);
+  ColumnDeviceView* h_column = reinterpret_cast<ColumnDeviceView*>(h_buffer.data());
+  // Each ColumnDeviceView instance may have child objects which may
+  // require setting some internal device pointers before being copied
+  // from CPU to device.
+  // Allocate the device memory to be used in the result.
+  // We need this pointer in order to pass it down when creating the
+  // ColumnDeviceViews so the column can set the pointer(s) for any
+  // of its child objects.
+  auto _descendant_storage = new rmm::device_buffer(views_size_bytes, stream);
+  auto _columns            = reinterpret_cast<ColumnDeviceView*>(_descendant_storage->data());
+  // The beginning of the memory must be the fixed-sized ColumnDeviceView
+  // objects in order for _columns to be used as an array. Therefore,
+  // any child data is assigned to the end of this array (h_end/d_end).
+  auto h_end = (int8_t*)(h_column + num_columns);
+  auto d_end = (int8_t*)(_columns + num_columns);
+  // Create the ColumnDeviceView from each column within the CPU memory
+  // Any column child data should be copied into h_end and any
+  // internal pointers should be set using d_end.
+  for (auto itr = source_view.begin(); itr != source_view.end(); ++itr) {
+    auto col = *itr;
+    // convert the ColumnView into ColumnDeviceView
+    new (h_column) ColumnDeviceView(col, h_end, d_end);
```

Comment:
This logic isn't doing anything to handle the alignment of the objects.

Reply:
That's a good point. This would also effect the size of the memory needed. So I recommend adding the following to **both** column device view `extent()` functions:
https://github.com/rapidsai/cudf/blob/35f9b234d350bb147dc188153d8b8f28c3e68fd0/cpp/src/column/column_device_view.cu#L172-L173
```
return cudf::util::round_up_safe(std::accumulate(
    get_extent, get_extent + source.num_children(), sizeof(column_device_view)),sizeof(column_device_view));
```
This should ensure all class pointers are aligned correctly.
(Don't forget the `mutable_column_device_view::extent()` function as well)

Reply:
See https://github.com/rapidsai/cudf/pull/6605#discussion_r514312391 for a more general solution to handling alignment.